### PR TITLE
@taskgroup decorator implementation  [https://github.com/apache/airflow/issues/11870 ]

### LINKED ISSUE #11870: Add @taskgroup decorator
**Description**

Similarly to `@dag` and `@task` decorators, we can add a decorator to easily generate `TaskGroup` instances. 

**Use case / motivation**

This would be used for reusable pieces of DAGs. An example may be to hyperparameter tune an ML pipeline. You may want to generate several parallel subpipelines with different settings/parameters to be executed in parallel. At the end pull the result and decide the best parametrization.  

```py
@task
collect_dataset(...)
  pass

@task
def train_model(...)
   pass

@task_group
def movielens_model_pipeline(learning_rate: int, feature_list: list, dataset: XComArg):  
  dataset = filter_dataset(dataset, feature_list)
  train_model(dataset, learning_rate)

@dag
def movielens_hpt(dataset_path: str, feature_list:list=['year', 'director']):
  dataset = load_dataset(dataset_path)
  for i in range(0.1, 0.9, step=0.1):
    movielens_model_pipeline(i, feature_list, dataset)
  
  decide_best_model(...)
```

**Related Issues**

https://github.com/apache/airflow/issues?q=is%3Aopen+is%3Aissue+label%3AAIP-31

### PR DESCRIPTION
**Pull request by VBhojawala**
New Feature of @taskgroup decorator implementation 
closes:#11870

Added a decorator function which can be used to create TaskGroup for python function.

I have tested it with following example : 

```py
def generate_value():
    """Dummy function"""
    return "Bring me a shrubbery!"


@task()
def print_value(value):
    """Dummy function"""
    ctx = get_current_context()
    log.info("Print Value: %s (at %s)", value, ctx['ts'])
    return value


@task()
def log_value(value):
    """Dummy function"""
    ctx = get_current_context()
    log.info("Log Value: %s (at %s)", value, ctx['ts'])


@taskgroup(group_id='t3_task_1')
def t3_task_1(value):
    op1 = print_value(value)
    log_value(op1)


with DAG(
    dag_id='T3_task_group_deco',
    default_args={'owner': 'airflow'},
    start_date=days_ago(2),
    schedule_interval=None,
    tags=['example'],
) as dag:
    task1 = PythonOperator(
        task_id='generate_value',
        python_callable=generate_value,
    )

    t3_task_1(task1.output)
```

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

**Comment by turbaszek** [comment]
CC @yuqian90

**Review by turbaszek — CHANGES_REQUESTED** [review]
Generally looks good! However, we will need some test and documentation + examples around that.

File: docs/concepts.rst [hunk 1]
```diff
@@ -1017,6 +1017,32 @@ This animated gif shows the UI interactions. TaskGroups are expanded or collapse
 
 .. image:: img/task_group.gif
 
+@taskgroup decorator
```
Comment (turbaszek) [inline docs/concepts.rst:1020]:
```suggestion
TaskGroup decorator
```

File: docs/concepts.rst [hunk 2]
```diff
@@ -1017,6 +1017,32 @@ This animated gif shows the UI interactions. TaskGroups are expanded or collapse
...
+--------------------
+
+TaskGroup can be created using @taskgroup decorator, it takes one mandatory argument 'group_id' which is same as constructor of TaskGroup class. It works exactly same as creating TaskGroup using context manager 'with TaskGroup('groupid') as section:'.
```
Comment (turbaszek) [inline docs/concepts.rst]:
```suggestion
TaskGroup can be created using ``@taskgroup decorator``, it takes one mandatory argument ``group_id`` which is same as constructor of TaskGroup class. It works exactly same as creating TaskGroup using context manager ``with TaskGroup('groupid') as section:``.
```

File: docs/concepts.rst [hunk 3]
```diff
@@ -1017,6 +1017,32 @@ This animated gif shows the UI interactions. TaskGroups are expanded or collapse
...
+.. code-block:: python
+
+  @task
+  def task_1(value):
+      return f'[ Task1 {value} ]'
+
+
+  @task
+  def task_2(value):
+      print( f'[ Task2 {value} ]')
```
Comment (turbaszek) [inline docs/concepts.rst]:
```suggestion
      print(f'[ Task2 {value} ]')
```

File: docs/concepts.rst [hunk 4]
```diff
@@ -1017,6 +1017,32 @@ This animated gif shows the UI interactions. TaskGroups are expanded or collapse
...
+
+
+  @taskgroup(group_id='section_1')
+  def section_1(value):
+      return task_2(task_1(value))
+
```
Comment (turbaszek) [inline docs/concepts.rst:1040]: I'm not sure if we need two examples. If we decide to keep both the it would be nice to add some text between them.

**Review by turbaszek — CHANGES_REQUESTED** [review]
@VBhojawala we still needs tests for this decorator to avoid regression in future

**Comment by yuqian90** [comment]
Thanks for proposing this. I think it works for small examples. However I have some questions regarding how this should work for larger examples. For instance, if we wanna put two or more tasks in the same `TaskGroup`, can that be done with with `@task_group`?

The reason I'm asking is that the contextmanager syntax is quite natural for creating nested structures (which is what `TaskGroup` is). The decorator syntax can probably be tweaked to do something close enough, but it's not as straightforward. If you have a good way of doing that, maybe demonstrate that in the example?

**Comment by VBhojawala** [comment]
> Thanks for proposing this. I think it works for small examples. However I have some questions regarding how this should work for larger examples. For instance, if we wanna put two or more tasks in the same `TaskGroup`, or a `TaskGroup` nested in another `TaskGroup`, can that be done with with `@task_group`?
> 
> The reason I'm asking is that the contextmanager syntax is quite natural for creating nested structures (which is what `TaskGroup` is). The decorator syntax can probably be tweaked to do something close enough, but it's not as straightforward. If you have a good way of doing that, maybe demonstrate that in the example?

Yes it does works because we have used the same context manager used TaskGroup, i am writing test cases for that also. Here is the example 


``` python
from airflow.models.dag import DAG
from airflow.operators.python import task
from airflow.utils.dates import days_ago
from airflow.utils.task_group import taskgroup


# Creating Tasks
from airflow.www.views import task_group_to_dict


@task()
def task_start():
    """Dummy Task which is First Task of Dag """
    return '[Task_start]'


@task()
def task_end1():
    """Dummy Task which is Last Task of Dag"""
    print(f'[ Task_End 1 ]')


@task()
def task_end2():
    """Dummy Task which is Last Task of Dag"""
    print(f'[ Task_End 2 ]')


@task
def task_1(value):
    """ Dummy Task1"""
    return f'[ Task1 {value} ]'


@task
def task_2(value):
    """ Dummy Task2"""
    print(f'[ Task2 {value} ]')


@task
def task_3(value):
    """ Dummy Task3"""
    return f'[ Task3 {value} ]'


@task
def task_4(value):
    """ Dummy Task3"""
    print(f'[ Task4 {value} ]')


# Creating TaskGroups
@taskgroup(group_id='section_1')
def section_1(value):
    """ TaskGroup for grouping related Tasks"""

    @taskgroup(group_id='section_2')
    def section_2(value2):
        """Nested TaskGroup for grouping related Tasks"""
        return task_4(task_3(value2))

    op1 = task_2(task_1(value))
    return section_2(op1)


# Executing Tasks and TaskGroups
with DAG(dag_id="example_nested_task_group_decorator", start_date=days_ago(2), tags=["example"]) as dag:
    t1 = task_start()
    s1 = section_1(t1)
    s1.set_downstream(task_end1())


```

**Comment by VBhojawala** [comment]
> > Thanks for proposing this. I think it works for small examples. However I have some questions regarding how this should work for larger examples. For instance, if we wanna put two or more tasks in the same `TaskGroup`, or a `TaskGroup` nested in another `TaskGroup`, can that be done with with `@task_group`?
> > The reason I'm asking is that the contextmanager syntax is quite natural for creating nested structures (which is what `TaskGroup` is). The decorator syntax can probably be tweaked to do something close enough, but it's not as straightforward. If you have a good way of doing that, maybe demonstrate that in the example?
> 
> Yes it does works because we have used the same context manager of TaskGroup, i am writing test cases for that also. Here is the example
> 
> ```python
> from airflow.models.dag import DAG
> from airflow.operators.python import task
> from airflow.utils.dates import days_ago
> from airflow.utils.task_group import taskgroup
> 
> 
> @task()
> def task_start():
>     """Dummy Task which is First Task of Dag """
>     return '[Task_start]'
> 
> 
> @task()
> def task_end():
>     """Dummy Task which is Last Task of Dag"""
>     print(f'[ Task_End ]')
> 
> 
> @task
> def task_1(value):
>     """ Dummy Task1"""
>     return f'[ Task1 {value} ]'
> 
> 
> @task
> def task_2(value):
>     """ Dummy Task2"""
>     print(f'[ Task2 {value} ]')
> 
> 
> @task
> def task_3(value):
>     """ Dummy Task3"""
>     return f'[ Task3 {value} ]'
> 
> 
> @task
> def task_4(value):
>     """ Dummy Task3"""
>     print(f'[ Task4 {value} ]')
> 
> 
> # Creating TaskGroups
> @taskgroup(group_id='section_1')
> def section_1(value):
>     """ TaskGroup for grouping related Tasks"""
> 
>     @taskgroup(group_id='section_2')
>     def section_2(value2):
>         """ Nested TaskGroup for grouping related Tasks"""
>         return task_4(task_3(value2))
> 
>     op1 = task_2(task_1(value))
>     return section_2(op1)
> 
> 
> with DAG(dag_id="example_nested_task_group_decorator", start_date=days_ago(2), tags=["example"]) as dag:
> 
>     t1 = task_start()
>     s1 = section_1(t1)
>     s1.set_downstream(task_end())
> ```

@turbaszek Should we include above example in example_dags folder?

**Comment by VBhojawala** [comment]
Hello @turbaszek,
I have  added test cases for taskgroup decorator. Kindly review it.

**Comment by VBhojawala** [comment]
To simplify the `@taskgroup ` decorator syntax, made group_id parameter optional . If taskgroup is only created with @taskgroup() decorator it will automatically assign group_id as function name.

**Comment by VBhojawala** [comment]
Re-raising pull request.
